exporter/awsxrayexporter/conn.go

```diff
@@ -0,0 +1,276 @@
+		if err != nil {
+			logger.Error("Unable to retrieve the region from the EC2 instance %v\n", zap.Error(err))
+		} else {
+			logger.Debug("Fetch region %v from ec2 metadata", zap.String("region", awsRegion))
+		}
+	}
+	if awsRegion == "" {
+		//log.Error("Cannot fetch region variable from config file, environment variables and ec2 metadata.")
+		os.Exit(1)
+	}
+	s = cn.newAWSSession(logger, cfg.RoleARN, awsRegion)
+
+	config := &aws.Config{
+		Region:                 aws.String(awsRegion),
+		DisableParamValidation: aws.Bool(true),
+		MaxRetries:             aws.Int(2),
+		Endpoint:               aws.String(cfg.Endpoint),
+		HTTPClient:             http,
+	}
+	return config, s
+}
+
+// ProxyServerTransport configures HTTP transport for TCP Proxy Server.
+func ProxyServerTransport(config *Config) *http.Transport {
+	tls := &tls.Config{
+		InsecureSkipVerify: config.NoVerifySSL,
+	}
+
+	proxyAddr := getProxyAddress(config.ProxyAddress)
+	proxyURL := getProxyURL(proxyAddr)
+
+	// Connection timeout in seconds
+	idleConnTimeout := time.Duration(config.RequestTimeout) * time.Second
+
+	transport := &http.Transport{
+		MaxIdleConns:        config.Concurrency,
+		MaxIdleConnsPerHost: config.Concurrency,
+		IdleConnTimeout:     idleConnTimeout,
+		Proxy:               http.ProxyURL(proxyURL),
+		TLSClientConfig:     tls,
+
+		// If not disabled the transport will add a gzip encoding header
+		// to requests with no `accept-encoding` header value. The header
+		// is added after we sign the request which invalidates the
+		// signature.
+		DisableCompression: true,
+	}
+
+	return transport
+}
+
+func (c *Conn) newAWSSession(logger *zap.Logger, roleArn string, region string) *session.Session {
+	var s *session.Session
+	var err error
+	if roleArn == "" {
+		s = getDefaultSession(logger)
+	} else {
+		stsCreds := getSTSCreds(logger, region, roleArn)
+
+		s, err = session.NewSession(&aws.Config{
+			Credentials: stsCreds,
+		})
+
+		if err != nil {
+			logger.Error("Error in creating session object : ", zap.Error(err))
+			os.Exit(1)
+		}
+	}
+	return s
+}
+
+// getSTSCreds gets STS credentials from regional endpoint. ErrCodeRegionDisabledException is received if the
+// STS regional endpoint is disabled. In this case STS credentials are fetched from STS primary regional endpoint
+// in the respective AWS partition.
+func getSTSCreds(logger *zap.Logger, region string, roleArn string) *credentials.Credentials {
+	t := getDefaultSession(logger)
+
+	stsCred := getSTSCredsFromRegionEndpoint(logger, t, region, roleArn)
+	// Make explicit call to fetch credentials.
+	_, err := stsCred.Get()
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case sts.ErrCodeRegionDisabledException:
+				logger.Error("Region : %v - %v", zap.String("region", region), zap.String("error", aerr.Error()))
+				logger.Info("Credentials for provided RoleARN will be fetched from STS primary region endpoint instead of regional endpoint.")
```

Comment:
Is there a need for both Error and Info? Would it be better to just do one Error logging?

Reply:
Fixed